app/services/dttp/contact_service/create.rb

```diff
@@ -0,0 +1,54 @@
+module Dttp
+  module ContactService
+    class Create
+      class << self
+        def call(**args)
+          new(**args).call
+        end
+      end
+
+      def initialize(trainee:)
+        @trainee = trainee
+      end
+
+      def call
+        Client.post(endpoint_path, {
+          body: body,
+          headers: headers,
+        })
+      end
+
+      private_class_method :new
+
+    private
+
+      attr_reader :trainee
+
+      delegate :first_names, :last_name, :dttp_id, to: :trainee
+
+      def auth_token
+        AccessTokenService.call
+      end
+
+      def headers
+        {
+          "Accept" => "application/json",
+          "Content-Type" => "application/json;odata.metadata=minimal",
+          "Authorization" => "Bearer #{auth_token}",
+        }
+      end
+
+      def body
+        {
+          "firstname" => first_names,
+          "lastname" => last_name,
+          "contactid" => dttp_id,
```

Comment:
is that correct? a `dttp_id` vs `contactid`

Reply:
It's `contactid` in DTTP but as far as the `Trainee` is concerned this is just its DTTP identifier. It's the unique ID in DTTP but we have to pass it in. I think all of our models that end up persisted to DTTP in some way will have one at which point we can extract it and share the setting, locking etc across the models.

Reply:
I prefer it generic as I don't think our models need to know which entity set they belong to etc. As far as the model is concerned it's just a UUID that can be used in 'other services'. In some ways the argument would be to not call it DTTP anything. If we leave it as `dttp_id` we can extract the code and reuse it in all of our other models so as the DTTP integration layer has a unique ID to refer to things by.

Reply:
As stated from [Section 1: Basic create and update operations](https://docs.microsoft.com/en-us/powerapps/developer/common-data-service/webapi/web-api-basic-operations-sample#section-1-basic-create-and-update-operations)

We should use the response header [OData-EntityId](https://docs.oasis-open.org/odata/odata/v4.0/os/part1-protocol/odata-v4.0-os-part1-protocol.html#_Toc372793637)

As the `id` generated from us may 
- increase fails (http status code `412`)
- further exacerbate fails with retries (compounding)
- conflicts with "single source of truth" and "system of record" 


The id & it's generation does not belongs here.

Reply:
@defong I've fixed this up now to use the returned entity id.